src/main.rs

```diff
@@ -0,0 +1,155 @@
+#![deny(unused_must_use)]
+#![deny(unsafe_code)]
+#![deny(clippy::unwrap_used)]
+
+use anyhow::Result;
+use dialoguer::{Confirm, Password, Select};
+use uom::si::thermodynamic_temperature::degree_celsius;
+use uom::si::Unit;
+
+use crate::cli::{prelude::*, Cli, Command};
+use crate::data::{Location, Provider, Weather};
+use crate::storage::Storage;
+use crate::ui::{eprintln, get_style_for_weather, println, sprintln, theme, with_spinner};
+
+mod api;
+mod cli;
+mod data;
+mod error;
+mod storage;
+mod ui;
+
+fn main() -> Result<()> {
+    human_panic::setup_panic!();
+    env_logger::init();
+    let args = Cli::parse();
+
+    match args.command {
+        Command::Configure { provider, config } => {
+            let mut storage = Storage::load(config.as_ref())?;
+            configure_provider(&mut storage, provider)?;
+            storage.store(config)?;
+        }
+        Command::Get {
+            provider,
+            location,
+            config,
+        } => {
+            let mut storage = Storage::load(config.as_ref())?;
+            let provider = choose_active_provider(&mut storage, provider);
+
+            let api_key = storage.get_api_key(provider).to_owned();
+            let api = api::new(provider, api_key);
+
+            let location = choose_location(&mut storage, api.as_ref(), location)?;
+            show_location(location);
+
+            let weather = with_spinner(|| api.get_weather(location))?;
+            show_weather(&weather);
+            storage.store(config)?;
+        }
+    }
+
+    Ok(())
+}
+
+fn configure_provider(storage: &mut Storage, provider: Provider) -> Result<()> {
+    if storage.is_provider_configured(provider) {
+        println("Provider is already configured.");
+        let confirmation = Confirm::with_theme(theme())
+            .with_prompt("Do you want to reconfigure?")
+            .interact()?;
+        if !confirmation {
+            println("Provider configuration has not changed.");
+            return Ok(());
+        }
+    }
+
+    let api_key: String = Password::with_theme(theme())
+        .with_prompt("Input provider API key")
+        .interact()?;
+    let api = api::new(provider, api_key.clone());
+
+    let is_correct_api_key = with_spinner(|| api.is_valid())?;
+    if !is_correct_api_key {
+        eprintln("Incorrect provider API key.")
```

Comment:
You can also print ```incorrect api key for provider: ...```
so when the person puts incorrect api key, It will know for which provider.
It can be helpful  IMO.

Reply:
Yes, I also was thinking about it, and even had such code some time, but after deleted it, because currently user always needs to specify manually provider name in the `configure` command, so it was redundant.